[ENH] Attempt to fix issues 737, 198, 752, 209, 366, 382, 695

Nalmeder opened the pull request:
<!-- Thank you for your PR!

BEFORE YOU CONTINUE! Please add the appropriate three-letter abbreviation to your title.

The abbreviations can be:
- [DOC]: Documentation fixes.
- [ENH]: Code contributions and new features.
- [TST]: Test-related contributions.
- [INF]: Infrastructure-related contributions.

Also, do not forget to tag the relevant issue here as well.

Finally, as commits come in, don't forget to regularly rebase!
-->

# PR Description

Please describe the changes proposed in the pull request:

- **We have added the enhancements listed in the above issues, as well as written tests and made sure the continuous integration passed and complied with all linters. Do note: there may be better solutions.**

<!-- Doing so provides maintainers with context on what the PR is, and can help us more effectively review your PR. -->

<!-- Please also identify below which issue that has been raised that you are going to close. -->

**This PR resolves #737 **
**This PR resolves #198 **
**This PR resolves #752 **
**This PR resolves #209 **
**This PR resolves #366 **
**This PR resolves #382 **
**This PR resolves #695 **

<!-- As you go down the PR template, please feel free to delete sections that are irrelevant. -->

# PR Checklist

<!-- This checklist exists for newcomers who are not yet familiar with our requirements. If you are experienced with
the project, please feel free to delete this section. -->

Please ensure that you have done the following:

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

<!-- Finally, please tag relevant maintainers to review. -->

Please tag maintainers to review.

- @ericmjl

Nalmeder committed:
Added line counter workflow

Nalmeder committed:
Update and rename python-app.yml to Line_check.yml

Nalmeder committed:
Testing to make sure we VCS works

Nalmeder committed:
Testing to make sure we VCS works

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
testing a thing

Sousa8697 committed:
Rough sketch of what I think we are trying to accomplish

Sousa8697 committed:
Wanted to include a small note regarding the appending of the row

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	janitor/functions.py

Nalmeder committed:
Conditional Join pseudocode is now working code that doesn't do anything yet

Sousa8697 committed:
fixing indexing issue

Sousa8697 committed:
Resolving appending issue with join_conditional function

Sousa8697 committed:
column name issue should be fixed now

Sousa8697 committed:
Added docstring documentation for function

Sousa8697 committed:
reformated docstring

Nalmeder committed:
trunc_datetime() method added.
TODO: docstring

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

MollyCroke committed:
initial effort to fix issue 752

MollyCroke committed:
fixes issue of space between "in" and "%" in column names not being replaced by underscore

Nalmeder committed:
Improved truncate, it can now be used more easily.

Nalmeder committed:
Made date truncation easier to use, added basic test

Nalmeder committed:
Made date truncation easier to use, added basic test

Nalmeder committed:
docstring cleanup to comply with flake8

Nalmeder committed:
Adjusted documentation to appeal to flake8

Nalmeder committed:
placate flake8

Nalmeder committed:
placate flake8

Nalmeder committed:
placate flake8

Nalmeder committed:
Attempt to comply with Black

Nalmeder committed:
Update pyproject.toml

Nalmeder committed:
Attempt to comply with Black

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
Code now complies with all linters

MollyCroke committed:
create test for issue 752

Sousa8697 committed:
Added a helper function and the remove_dupes function. Still need to write tests.

Sousa8697 committed:
Merge remote-tracking branch 'origin/dev' into dev

MollyCroke committed:
Add more tests for issue 752

MollyCroke committed:
Update functions.py and fix tests

Nalmeder committed:
improved trunc_datetime

Nalmeder committed:
improved test_trunc_datetime

Nalmeder committed:
linter compliance

Nalmeder committed:
linter compliance

Sousa8697 committed:
Reformatted files, added tests for remove_dupes

Nalmeder committed:
improved tests

Sousa8697 committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	janitor/functions.py

Sousa8697 committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
Linter compliance

Sousa8697 committed:
Reformatted files, added exception handling to remove_dupes, added tests for remove_dupes

Sousa8697 committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
Linter compliance

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
Linter compliance

Nalmeder committed:
Linter compliance

Nalmeder committed:
Update test_remove_dupes.py

Nalmeder committed:
Update conftest.py

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
Update test_remove_dupes.py

Sousa8697 committed:
changed test_remove_dupes to fix error

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

Sousa8697 committed:
changed test_remove_dupes to fix error

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
Linter compliance

Nalmeder committed:
Better testing for join_conditional

Nalmeder committed:
linter compliance

Sousa8697 committed:
Added sorting function based on column values order

Sousa8697 committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
Merge remote-tracking branch 'origin/dev' into dev

Nalmeder committed:
linter compliance

Nalmeder committed:
linter compliance

ericlessantostv committed:
Added new functions

Added convert_stock, is_connected, and get_symbol functions to functions.py

ericlessantostv committed:
fixed formatting of functions in functions.py

ericlessantostv committed:
Create test_get_symbol.py

ericlessantostv committed:
Create test_is_connected.py

MollyCroke committed:
Update tests

MollyCroke committed:
Merge remote-tracking branch 'origin/dev' into dev

ericlessantostv committed:
Create test_convert_stock.py

ericlessantostv committed:
Create test_sort_column_value_order.py

ericlessantostv committed:
Update test_sort_column_value_order.py

ericlessantostv committed:
Update test_sort_column_value_order.py

ericlessantostv committed:
fixed imports

ericlessantostv committed:
Update test_get_symbol.py

ericlessantostv committed:
Update test_is_connected.py

ericlessantostv committed:
Update test_convert_stock.py

Nalmeder committed:
linter compliance

Nalmeder committed:
linter compliance

Nalmeder committed:
linter compliance

Nalmeder committed:
linter compliance

Nalmeder committed:
linter compliance

Nalmeder committed:
linter compliance

Nalmeder committed:
linter compliance

Nalmeder committed:
Literally added a period to make the test work

Nalmeder committed:
lowercase f

Nalmeder committed:
test improvements

Nalmeder committed:
Update AUTHORS.rst

ericmjl commented:
Hi @BaritoneBeard! Thank you for your PR. I appreciate the effort!

At the same time, the PR is quite a big one, so this might take a bit of time to review. It'd be awesome to knock out a bunch of issues in the same PR, though, so I'm happy to give this a review. :smile:

Would you be open to:

1. Fixing up the conflicts,
2. Ensuring all tests pass on the PR branch,
3. Ensuring test coverage is maintained or improved?

Once those are fulfilled, I'm happy to take some time to do a more in-depth review.

Please let me know if you have questions!

Nalmeder committed:
Merge branch 'dev' into dev

Nalmeder commented:
@ericmjl 
Absolutely, I have resolved the conflicts and the tests/linting is all green on the PR branch.

ericmjl commented:
@BaritoneBeard thanks a ton! I just approved the PR checks run. I'm going to ping the team to see how we can divvy up the review labour, hopefully to make the process run a bit faster. Hang tight!

ericmjl commented on .github/workflows/Line_check.yml:
```diff
@@ -0,0 +1,25 @@
+# This workflow will install Python dependencies, run tests and lint with a single version of Python
```
I think this file should be removed, as we have an official suite of tests that will be run.

ericmjl replied:
**Action item:** Please remove this file.

ericmjl commented on janitor/functions.py:
```diff
@@ -55,6 +57,78 @@
 )
 
 
+def get_symbol(symbol: str) -> str:
```
I think this function might be better suited inside the `finance.py` submodule. Would you be open to moving it there?

ericmjl replied:
**Action item:** Please chime in on your thoughts about moving the function to the `finance` submodule, and if the answer is affirmative, please go ahead.

ericmjl commented on janitor/functions.py:
```diff
@@ -55,6 +57,78 @@
+            return "Not found."
+
+
+def is_connected(url: str) -> bool:
```
This is quite a nice function! Thanks for adding it in.

That said, I think it might be better placed in `utils.py`. What are your thoughts here?

ericmjl replied:
The more I think about it, the more I am convinced this is a utility function rather than a main data cleaning function.

**Action item:** Please move the function to `utils.py`.

ericmjl commented on janitor/functions.py:
```diff
@@ -55,6 +57,78 @@
+        if sock is not None:
+            sock.close()
+            return True
+    except OSError:
+        pass
```
I think I'm lacking some context. Under what circumstances will an OSError show up?

Following Pythonic practice, I might be better to raise the error explicitly and loudly rather than let it silently pass. What do you think of the following pattern instead?

```suggestion
    except OSError as e:
        import warnings
        warnings.warn("There was an issue connecting to the internet. Please see original error below.")
        raise e
```

ericmjl replied:
**Action item:** Please address the questions above. We will discuss further on what the most optimal pattern here ought to look like.

ericmjl commented on janitor/functions.py:
```diff
@@ -1811,6 +1888,81 @@ def remove_columns(
+    return df
+
+
+def remove_dupes(df: pd.DataFrame, keep="first") -> pd.DataFrame:
```
I don't mean here to undo work already done, but I think we might not need this function, as the exact functionality desired is available in `pandas` native itself. (I checked on [the docs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html).) Definitely this is my fault for not keeping the issue tracker refreshed; #209 should have been closed a long time ago.

ericmjl replied:
**Action item:** Please remove the function `remove_dupes` and the helper function `get_occurrences`. The functionality is already covered in `pandas`.

ericmjl commented on janitor/functions.py:
```diff
@@ -3657,6 +3809,63 @@ def to_datetime(
 
 
+@pf.register_dataframe_method
+def trunc_datetime(datepart: str, timestamp: dt.datetime):
```
Because this function is registered as a dataframe method through the `register_dataframe_method` decorator, it is missing the first argument `df`, and by extension, the column name on which this operation should be performed.

ericmjl replied:
On second look, I think this function is actually an "elementary function" (akin to the NumPy universal functions) that operate elementwise. 

Looking at the function body, I think the function as it stands is not yet a dataframe method. 

**Action items:** Please do the following:

1. Rename function to `truncate_datetime`.
2. Implement it such that it works with the pattern `df.truncate_datetime(...)`
3. Add a test for this one.

ericmjl commented on janitor/functions.py:
```diff
@@ -4055,6 +4264,89 @@ def take_a_if_even(x):
+
+    This method does not change the original DataFrame.
+
+     This function is supposed to emulate SQL and R's
```
From this line onwards, the indentation appears to be off. Can you make sure the function docstring's indentation is consistent?

ericmjl replied:
**Action item:** Please ensure indentation is correct. At the end, please also check that the function documentation is visible on the docs preview. Ping us if you are unsure how to check the docs preview! :smile:

ericmjl commented on janitor/functions.py:
```diff
@@ -4055,6 +4264,89 @@ def take_a_if_even(x):
+    :param join_operator: A string that represents the operator
+         in the comparison e.g. < in the above ex.
+    :returns: A pandas DataFrame with rows that matched our comparison.
+
+
```
(nitpick) Extraneous blank lines. Could you remove these please?

ericmjl replied:
**Action item:** Please remove these extraneous blank lines.

ericmjl commented on janitor/functions.py:
```diff
@@ -4055,6 +4264,89 @@ def take_a_if_even(x):
+    return_df = pd.DataFrame(columns=df.columns.tolist())
+    curr_in = 0
+    for index, row in df.iterrows():
+        if join_operator == ">":
+            if row[left_column] > row[right_column]:
+                return_df.loc[curr_in] = list(df.loc[index])
+                curr_in += 1
+        elif join_operator == "<":
+            if row[left_column] < row[right_column]:
+                return_df.loc[curr_in] = list(df.loc[index])
+                curr_in += 1
+        elif join_operator == "!=":
+            if row[left_column] != row[right_column]:
+                return_df.loc[curr_in] = list(df.loc[index])
+                curr_in += 1
```
This is more of a note to myself than to you, @BaritoneBeard, but I think there is a cleaner way to handle this chunk of code without the if/elif/else pattern and code duplication that I see here. While this pattern is much more explicit, it also involves more duplication, which makes extending the function with other operators a bit more challenging.

ericmjl commented on janitor/functions.py:
```diff
@@ -4434,6 +4726,68 @@ def sort_naturally(
+                new_df = df.sort_values(columns + ["cond_order"])
+                del new_df["cond_order"]
+            return new_df
+        else:
+            print("ValueError: Column Name not in DataFrame")
+            return pd.DataFrame()
+    else:
+        print("ValueError: column_value_order dictionary cannot be empty")
+        return pd.DataFrame()
```
Errors should be raised, rather than printed. Could you please change these lines to raise an error instead of printing it?

ericmjl replied:
**Action item:** As per comment above.

ericmjl commented on tests/functions/test_convert_stock.py:
```diff
@@ -0,0 +1,7 @@
+
+
+def test_convert_stock():
+    assert get_symbol("GME") == "GameStop Corp."
```
Nice contemporary culture reference. 😎

ericmjl commented on tests/functions/test_convert_stock.py:
```diff
@@ -0,0 +1,7 @@
+from janitor import get_symbol
```
Because this is a test for a finance-themed function, I think it should go under the finance tests instead.

ericmjl replied:
**Action item:** Please move this into the finance.py submodule.

ericmjl commented on tests/functions/test_convert_stock.py:
```diff
@@ -0,0 +1,7 @@
+from janitor import get_symbol
+
+
+def test_convert_stock():
```
Could you add a docstring here too, please? The intent behind the test should be documented. 

I can foresee one failure mode for this test - if the `GME` symbol gets assigned to a different name, then the test will fail. Can you add to the docstring to which website we should go to see what the expected return string ought to be? That would go a long way to ensuring long-term maintainability of the function.

ericmjl replied:
**Action item:** As per comment above.

ericmjl commented on tests/functions/test_get_symbol.py:
```diff
@@ -0,0 +1,8 @@
+from janitor import get_symbol
```
As with above, I think this function should be in the `finance` tests.

ericmjl replied:
**Action item:** As per above.

ericmjl commented on tests/functions/test_is_connected.py:
```diff
@@ -0,0 +1,8 @@
+from janitor.functions import is_connected
+
+
+def test_is_connected():
```
As with above, could we get a docstring added here please?

ericmjl replied:
**Action item:** As per comment above.

ericmjl commented on tests/functions/test_get_symbol.py:
```diff
@@ -0,0 +1,8 @@
+from janitor import get_symbol
+
+
+def test_get_symbol():
```
As with the above comments, could we get a docstring added here please?

ericmjl replied:
**Action item:** As per above.

ericmjl commented on tests/functions/test_is_connected.py:
```diff
@@ -0,0 +1,8 @@
+from janitor.functions import is_connected
```
If you move the `is_connected` function to `utils.py`, could you also move this corresponding test too?

ericmjl replied:
**Action item:** As per comment above.

ericmjl commented on tests/functions/test_join_conditional.py:
```diff
@@ -0,0 +1,26 @@
+
+
+@pytest.mark.functions
+def test_join_conditional():
```
Could we get a docstring here, while also including in there that this is an example-based test?

ericmjl replied:
**Action item:** As per title.

ericmjl commented on tests/functions/test_nonstandard_charac.py:
```diff
@@ -0,0 +1,26 @@
+import pandas as pd
```
It appears that the tests here are for the function `clean_names`. Could you move them into the `test_clean_names.py` file?

ericmjl replied:
**Action item:** As per comment above.

ericmjl commented on tests/functions/test_nonstandard_charac.py:
```diff
@@ -0,0 +1,26 @@
+
+
+@pytest.mark.functions
+def test_charac():
```
Could we also get a docstring here, please?

ericmjl replied:
**Action item:** As per comment above.

ericmjl commented on tests/functions/test_nonstandard_charac.py:
```diff
@@ -0,0 +1,26 @@
+    assert "current_accountbalance_in_%_of_gdp" in df.columns.values
+
+
+def test_space():
```
Same here, docstring please!

ericmjl replied:
**Action item:** As per comment above.

ericmjl commented on tests/functions/test_remove_dupes.py:
```diff
@@ -0,0 +1,35 @@
+
+@pytest.mark.functions
+def test_remove_dupes():
+    """
+    Tests for remove_dupes():
+    Test 1: passing in an object that is not a Pandas DataFrame should
+    return an empty DataFrame
+    Test 2: passing in a value that is not last or first will return the
+    DataFrame using "first" by default
+    Test 3: passing in an empty DataFrame should return an empty DataFrame
+    """
```
This is a great example of documenting the intent of the test!

ericmjl commented on tests/functions/test_sort_column_value_order.py:
```diff
@@ -0,0 +1,35 @@
+from janitor.functions import sort_column_value_order
+
+
+def test_sort_column_value_order():
```
Docstring please!

ericmjl replied:
**Action item:** As per comment above.

ericmjl commented on tests/functions/test_trunc_datetime.py:
```diff
@@ -0,0 +1,27 @@
+
+
+@pytest.mark.functions
+def test_trunc_datetime():
```
Docstring please!

ericmjl replied:
**Action item:** As per comment above.

ericmjl reviewed (changes requested):
@BaritoneBeard, thanks for the PRs from you and your teammates :smile:. 

I've added in my review, though I'd encourage you all to wait maybe a day or two more before taking action on it, so that others on the dev team can chime in.

Can I ask, how are you all (@Sousa8697 @MollyCroke and @ericclessantostv and yourself) related professionally? Just curious if this PR you all submitted together is part of a class project or team project of some kind?

Nalmeder commented:
@ericmjl

> Can I ask, how are you all (@Sousa8697 @MollyCroke and @ericclessantostv and yourself) related professionally? Just curious if this PR you all submitted together is part of a class project or team project of some kind?

 Sure, I don't mind disclosing that we just finished our undergrad yesterday. Our final assignment was to find a project on github, fork it, and fix open issues with it, making sure to practice clean coding, team working, continuous integration, the works. We've all taken stats classes and I also have a math degree so we were familiar with general data science, R, and SQLite, thus choosing pyjanitor was kind of a no brainer.

We already passed the class, but I figured we should submit the PR since we put the effort in. Worst case it gets denied, best case we helped out the project in some way.

ericmjl commented:
@BaritoneBeard that's a wonderful experience for you all to have! I kind of wish we had the same experience too.

Since you all put in the work and there's some good stuff in the PR too, let's make sure we take it to completion. :smile:

Congratulations on finishing your undergrad studies, btw!

Nalmeder commented:
@ericmjl 
> Congratulations on completing your undergrad studies

thank you very much!


> lets make sure we take it to completion

Ill do as you said earlier and give it a day or two for the rest of the change requests to roll in and then get to work on them.

hectormz commented:
Thanks for the epic PR! I haven't looked through yet, but does this satisfy all docstrings for #695 or only in the files changed? Just a note to keep in mind for fellow reviewers.

ericmjl commented on janitor/functions.py:
```diff
@@ -4055,6 +4264,89 @@ def take_a_if_even(x):
 
 
+@pf.register_dataframe_method
+def join_conditional(
```
@hectormz could I assign this one to you to review, please?

hectormz replied:
Great!

hectormz replied:
I don't think this `join_conditional` accomplishes what #737 is suggesting. If I read this correctly, this is performing conditional comparisons on rows in the same dataframe. But this should be performed on _two_ different dataframes. Do I have that right @BaritoneBeard and @samukweku ?

samukweku replied:
@hectormz @BaritoneBeard @ericmjl You should be able to use it on the same dataframe, like a self join. However, this does not match the expected output, as every value in one column should be compared to every other value in the other column. This is achievable with a cartesian product, which is memory intensive at least. So, for the example used in the test, the output should be : 

```python
  SalesMonth  Company1  Company2
0        Jan     150.0     180.0
1        Jan     150.0     250.0
2        Jan     150.0     500.0
3        Feb     200.0     250.0
4        Feb     200.0     500.0
5        Mar     300.0     500.0
6      April     400.0     500.0
```

where every value in `company1` column is compared against every value in `company2` columnn, and only the values where `company1` is less than `company2` is returned.

ericmjl replied:
That's a good point, @samukweku. @BaritoneBeard, please see the follow-up thread below for the action item.

jk3587 commented on janitor/functions.py:
```diff
@@ -55,6 +57,78 @@
+        return get_symbol(stock_symbol.upper())
+    else:
+        print("Connection Error: Client Not Connected to Internet")
+        return ""
```
Should raise exception here instead of returning empty string.

ericmjl replied:
Agree with @jk3587.

**Action item:** Please raise a `ConnectionError` that raises the error message that you currently have inside the print function. `ConnectionError` is an appropriate [built-in exception](https://docs.python.org/3/library/exceptions.html) that you can use.

hectormz commented on janitor/functions.py:
```diff
@@ -7,6 +7,8 @@
 import unicodedata
 import warnings
 from functools import partial, reduce
+import requests
```
Was `isort` run against the code (followed by `black`)? I think these imports should be arranged a little differently

ericmjl replied:
**Action item:** Please apply `isort` followed by `black` on the entire library before the final commit.

jk3587 commented on janitor/functions.py:
```diff
@@ -55,6 +57,78 @@
+    for x in result["ResultSet"]["Result"]:
+        if x["symbol"] == symbol:
+            return x["name"]
+        else:
+            return "Not found."
```
This makes it confusing when you can either return None or "Not Found". I think if no matches are found either due to 0 results or due to not matching any of the results, there should be a unified return (I prefer returning None). This way your test cases for ASNF and ZZZZ will return the same thing since there were no matches.

ericmjl replied:
I agree with @jk3587 here. There are some advantages, as it makes the return type compatible with pandas' `df.dropna()` function.

**Action item:** Please change the return to None.

ericmjl commented on janitor/functions.py:
```diff
@@ -3657,6 +3809,63 @@ def to_datetime(
+        4: timestamp.minute,
+        5: timestamp.second,
+    }
+    try:
+        ENUM[datepart]
+    except KeyError:
+        print("Invalid truncation")
+        print("Please enter Year, Month, Day, Hour, Minute, or Second")
+        return
```
For this block, I think the Pythonic pattern is to raise the error loudly and interrupt the program rather than letting it pass through. Here's a suggested modification that would bring this in-line:

```suggestion
    try:
        ENUM[datepart]
    # Capture the error but replace it with explicit instructions.
    except KeyError:
        msg = "Invalid truncation. Please enter any one of 'year', 'month', 'day', 'hour', 'minute' or 'second'."
        raise KeyError(msg)
```

ericmjl replied:
**Action Item:** Please change this block according to the suggestion above.

ericmjl commented on janitor/functions.py:
```diff
@@ -3657,6 +3809,63 @@ def to_datetime(
+        return
+
+    for i in range(ENUM.get(datepart) + 1):
+        print(ENUM.get(i))
```
This looks like a stray print statement. It should be removed.

ericmjl replied:
**Action Item:** Please remove the stray print statement.

ericmjl commented on janitor/functions.py:
```diff
@@ -4055,6 +4264,89 @@ def take_a_if_even(x):
 
 
+@pf.register_dataframe_method
+def join_conditional(
```
@BaritoneBeard this is my own review; we can wait on @hectormz's as well, but I think the function here doesn't implement `join_conditional`, and its implementation should be changed to implement it correctly.

What's implemented in the function right now is instead a query operation, which is trivially done using `pandas`' existing `.query()` API. For example, one may do:

```python
df.query("left_column < right_column")
```

and we'll be done.

I think what's intended by `join_conditional` is filtering when joining two tables. Looking around online, I see a lot of discussion in the `pandas` dev chatter:

- https://github.com/pandas-dev/pandas/issues/8962
- https://github.com/pandas-dev/pandas/issues/10309

On studying those implementations, I think what's intended is the following implementation:

```python
@pf.register_dataframe_method
def join_conditional(df: pd.DataFrame, other: Union[pd.DataFrame, None], conditions: List[str], join_kwargs={}):
    """Conditional join between one dataframe and another.

    This is syntactic sugar for joining (not merging) two dataframes and then filtering the results.
    As such, a strong assumption here is that the indexes
    are semantically compatible between the two dataframes.

    The conditions that are passed in have to be compatible with dataframe query strings.
    For example, if you want to state that `one_column` must be less than `another_column`,
    then your condition string should be `"one_columns < another_column"`.

    Multiple conditions are allowed. 
    """
    if other is None:
        other = df
    joined_df = df.join(other, **join_kwargs)
    for condition in conditions:
        joined_df = joined_df.query(condition)
    return joined_df
```

hectormz replied:
@ericmjl I agree. I see that my comments from last night didn't get posted. Whoops! But thanks for putting in the proposed change too

ericmjl replied:
Addressing @samukweku's point, self `join`s are supported by allowing `other` to be None in the edited implementation above.

**Action Items:**

1. Please change the implementation to match what is proposed above.
2. Please ensure the docstring is fully fleshed out like the other functions in the library. (I only put in the not-so-obvious points that a user might need to be reminded of.)

ericmjl reviewed (changes requested):
I had a chunk of time to do a second pass review this morning. It appears that `join_conditional` isn't implemented the way most people might understand it. @hectormz, could you help me double-check please?

ericmjl commented:
@BaritoneBeard - heads-up, I'm going to do a final pass on the PR and mark out what the action items are from each reviewer comments. Once that's done, please go ahead with the changes!

ericmjl reviewed (changes requested):
@BaritoneBeard, final review pass is done from my side! Please let us know if you have questions. Looking forward to doing a final review; please don't resolve any of the conversations, as I'll be using them as markers. If you would like to use them as markers for yourself, please feel free to leave comments like "Done!" for yourself.

samukweku commented:
@BaritoneBeard just to reiterate that the conditional compares every value in one column against every value in the other column. Don't think of it as a per row comparison

Nalmeder committed:
Update janitor/functions.py

Co-authored-by: Eric Ma <ericmjl@users.noreply.github.com>

Nalmeder committed:
Update janitor/functions.py

Co-authored-by: Eric Ma <ericmjl@users.noreply.github.com>

Nalmeder committed:
Removed line_check.yml file

As per request of moderator

Nalmeder committed:
moved test_convert_stock.py from functions to finance

Nalmeder committed:
Updated and moved test_get_symbol.py to finance

Nalmeder committed:
Update test_convert_stock.py

Nalmeder committed:
Update test_get_symbol.py

Nalmeder committed:
Updated and moved test_is_connected.py to utils

Nalmeder commented:
Hey @ericmjl 

I got very sick so I wasn't able to start working on the requested changes as early as I planned to, but I'm here now, and I'm working my way through them.

Because I was not involved with the join_conditional I actually have no clue whats going on there, We split the work up rather evenly so I only had a hand in about a quarter of this PR and I'm trying to piece together what I can for now. I think I am going to nix the join_conditional from the PR and maybe come back to it at a different time. I plan to adjust the datetime truncation method to include other styles of dates at some point in the future (for example, EU datetimes are usually Day/Month/Year, while the US writes it Month/Day/Year) so at that time I'll likely come back to it. What are your thoughts on this?

Other than that, I'll keep trucking on the changes requested.

ericmjl commented:
Hey @BaritoneBeard!

> I got very sick so I wasn't able to start working on the requested changes as early as I planned to, but I'm here now, and I'm working my way through them.

Please take your time! There's no pressure here :smile:. If you need more rest, you should take it. You shouldn't feel bad for dropping OSS work; that'll just take all the fun out of it.

> Because I was not involved with the join_conditional I actually have no clue whats going on there, We split the work up rather evenly so I only had a hand in about a quarter of this PR and I'm trying to piece together what I can for now. I think I am going to nix the join_conditional from the PR and maybe come back to it at a different time. 

Understood; I think that's a good plan going forward. Perhaps you might want to encourage the classmate who did this function to submit their own PR? That'll ensure their contribution gets officially recorded in the git history when we merge.

> I plan to adjust the datetime truncation method to include other styles of dates at some point in the future (for example, EU datetimes are usually Day/Month/Year, while the US writes it Month/Day/Year) so at that time I'll likely come back to it. What are your thoughts on this?

Your proposal sounds great. It's much better for PRs to be small in scope, as it makes things easier to review, as I mentioned before. What you have right now is already pretty good.

Nalmeder committed:
Update test_clean_names.py

added additional tests for nonstandard characters

Nalmeder committed:
Delete test_nonstandard_charac.py

There was no nonstandard_charac function to test, it was an addendum to clean_names

Nalmeder committed:
Update test_sort_column_value_order.py

Nalmeder committed:
Update test_trunc_datetime.py

Nalmeder committed:
isort, black

Nalmeder committed:
Various changes
 - Removed join_conditional
 - moved is_connected and get_symbol to utils and finance
 - many code cleanups/corrections

Nalmeder committed:
isort, black, renamed trunc_datetime

Nalmeder committed:
renamed appropriate test functions

Nalmeder committed:
flake8 compliance

Nalmeder committed:
isort, black

Nalmeder committed:
isort, black

Nalmeder committed:
linter compliance

Nalmeder committed:
Fixed imports, removed deleted test function

Nalmeder committed:
Fixed imports, removed deleted test function

Nalmeder committed:
Fixed imports, removed deleted test function

Nalmeder committed:
Fixed some tests

ericmjl commented:
@BaritoneBeard it appears there's a conflict. I'm going to resolve it, but you might need to pull down changes before pushing any more changes that you make.

ericmjl committed:
Merge branch 'dev' into dev

ericmjl commented:
@BaritoneBeard how are we doing here? Is there anything we could do to push this PR along? Most of what's failing is just the code style checks; everything else appears to be working fine! :smile:

Nalmeder commented:
@ericmjl 
Yeah, sorry about that I planned to work on it but my job has been demanding my free time lately. I plan to work on it on my lunch breaks this week so hopefully everything should be good. 

I just need to alter truncate_datetime and then I should be done on the code side.

Also, for code style: darglint in particular was giving me trouble last I worked on this and I couldn't figure out what I was doing differently so I may be pinging you in the future for some assitance. First I'll work on the code, then fix the style when thats done.

ericmjl commented:
Awesomeness, @BaritoneBeard! Thanks for letting us know. No pressure, btw, just wanted to make sure the PR didn't get dropped, that's all, especially given the good stuff that's in there. Do make sure you enjoy your lunch breaks too! It's always important to get some rest in between.

Nalmeder committed:
Adjusted Truncate Datetime method, added a df method to truncate throughout entire df

Nalmeder committed:
Linter compliance

Nalmeder committed:
Linter compliance

Nalmeder committed:
Linter compliance

ericmjl commented:
@BaritoneBeard are you encountering issues with `black`? If so, let me know if you'd like me to jump on a call with you. I recognize that satisfying a linter is one of the most annoying parts of software development and that working with someone through it is the best way to solve it.

Nalmeder commented:
Alrighty @ericmjl It seems my code stuff is finished, all thats left are some Linter compliance problems that I can't quite figure out, hoping you or another mod may be able to help out.

1) After running black, my terminal tells me utils.py has been reformatted, upon committing changes, it states everything is up to date. On push, black fails. Thus it seems I'm in a strange loop where I can't format utils.py and thus the black compliance check will always fail.

2) I have no idea what darglint wants, I can't quite see what I'm doing differently from other functions, and my initial PR had no darglint errors, so it's not like my group and I did something particularly wrong the first time. This is probably just due to my lack of familiarity with it.

I see you've asked a question before I was able to comment. If a call is best for you that works for me, we can decide a date and time. If you've seen these particular errors before and know a quick solution you can just let me know and I can work with it some more. Whatever's easiest.

ericmjl commented:
@BaritoneBeard definitely, let's schedule a call. Can you send me an email via http://shortwhale.com/ericmjl? (I use Shortwhale to protect my email address from the public.)

Nalmeder commented:
Hey @ericmjl I sent you an email a few weeks ago and haven't heard from you, no rush of course I just wanted to make sure you received it because if not I can simply send another. If you have received it an you're just busy, no worries I will continue to chill.

ericmjl commented:
Hey @BaritoneBeard! I searched in my Inbox and couldn't find any email from you. No worries, these things happen. Could you send it again?

Nalmeder commented:
@ericmjl Sure thing, it's been sent just now so you should have it at this time

ericmjl commented:
@BaritoneBeard this is weird, I still haven't seen anything in my inbox sent from you via Shortwhale.

No worries, can you send an email to me directly?

Really sorry we ended up going back and forth on such a trivial matter. I'll want to see whether Shortwhale is working for you when we call too, so I can make a choice on whether I want to keep using it or not.

Nalmeder committed:
Linter compliance

ericmjl committed:
Merge branch 'dev' into dev

Nalmeder committed:
Linter compliance

Nalmeder committed:
Linter compliance

Nalmeder committed:
Linter compliance

Nalmeder committed:
Linter compliance

Nalmeder committed:
Linter compliance

Nalmeder committed:
Commented Cat Coverage.xml

ericmjl commented:
Woohoo, we're all good! :smile:

ericmjl commented:
Writing down the context for future reference: @BaritoneBeard and I did a video call this evening to resolve issues that were preventing the PR checks from passing.
